include/picongpu/plugins/saxs/Saxs.hpp

```diff
@@ -0,0 +1,813 @@
+            constexpr uint32_t numWorkers = pmacc::traits::GetNumWorkers<
+                pmacc::math::CT::volume< SuperCellSize >::type::value >::value;
+
+            // Form factor density:
+            // Get the offset to the local domain (this HOST + DEVICE pair).
+            const SubGrid< simDim > & subGrid =
+                Environment< simDim >::get( ).SubGrid( );
+            DataSpace<simDim> globalOffset( subGrid.getLocalDomain( ).offset );
+            // Calculate the density and get a data box to access this TmpField.
+            DataConnector & dc = Environment<>::get( ).DataConnector( );
+            FieldTmp::DataBoxType tmpFieldBox = calculateDensity(
+                dc,
+                globalOffset
+            );
+            // Get the in cell position of a TmpField.
+            // Could probably remove it as it is the cell origin in all cell
+            // types.
+            const picongpu::traits::FieldPosition<
+                typename fields::CellType,
+                FieldTmp
+            > fieldPos;
+            // Shift the density box to exclude the GUARD.
+            DataSpace < simDim > guardingSC =
+                cellDescription->getGuardingSuperCells( );
+            auto const fieldTmpNoGuard =
+                tmpFieldBox.shift( guardingSC * SuperCellSize::toRT( ) );
+            // Get the field size on this rank (no GUARD).
+            DataSpace < simDim > cellsGrid =
+                ( cellDescription->getGridSuperCells( )  - 2 * guardingSC ) *
+                SuperCellSize::toRT( );
+            uint32_t const totalNumCells = cellsGrid.productOfComponents( );
+            // Get the number of, virtual, blocks.
+            PMACC_ASSERT( totalNumCells % numWorkers == 0 );
+            uint32_t const numBlocks = totalNumCells / numWorkers;
+
+
+            // Run Kernel.
+            if ( outputLayout == OutputMemoryLayout::SplitOverNodes )
+            {
+                runKernelInSplitMode(
+                    cellsGrid,
+                    fieldTmpNoGuard,
+                    globalOffset,
+                    numBlocks,
+                    fieldPos
+                );
+            }
+            else
+            {
+                runKernelInMirrorMode(
+                    cellsGrid,
+                    fieldTmpNoGuard,
+                    globalOffset,
+                    numBlocks,
+                    fieldPos
+                );
+            }
+            // Release density data.
+            dc.releaseData( FieldTmp::getUniqueId( 0 ) );
+            // Write to disk.
+            if ( pluginSystem::containsStep( outputPeriod, currentStep ) )
+                writeOutput( );
+            // Update the total number of rotations ( data passes ).
+            if ( outputLayout == OutputMemoryLayout::SplitOverNodes )
+                accumulatedRotations += countRanks - 1;
+        }
+    };
+} // end saxs
+} // end plugins
+namespace particles
+{
+namespace traits
```

Comment:
It is defined in the same way at the end of `Radiation.hpp`. To be honest I just copypasted it and adjusted. We could define it somewhere in `picongpu/particles/traits` but it doesn't seem to be the usual way to do it.

Reply:
@psychocoderHPC @sbastrakov What is your opinion?

Reply:
(I see this part was changed since, but logically it is similar).
I am actually not sure if this should logically be a trait. It depends on whether the plugin allows customizing it externally (i.e. from .param files). Ofc this is the question for @pordyna as a developer. In case such customization is not intended, I think it is better to have just a using-alias. In case the trait makes sense, I would prefer have SAXS-specific ones here and general ones (such as IsIon) perhaps with general PIConGPU particle traits.

Reply:
Ohh I actually understood the question is about the `SpeciesEligibleForSolver` specialization for the plugin. 
Thanks @sbastrakov for the explanation :smile: . I don't think that any of the "traits" in `.../saxs/traits` are meant to be a part of the public interface. For example `IsIon` is basically just checking if the species has `boundElectrons` but this definition of an ion will not be necessarily appropriate for a different plugin/feature.  So I think we can just remove the `saxs::traits` namespace.

Reply:
Ah, I might have also misunderstood. So just to clarify: my previous post concerned the new traits that your plugin creates, and perhaps these should not be traits based on your reply. Ofc you are free to use PIConGPU general traits like `SpeciesEligibleForSolver` when necessary.